app/src/main/java/com/timothyolt/evolutionarydesign/image/ImageViewHolder.kt

```diff
@@ -0,0 +1,52 @@
+package com.timothyolt.evolutionarydesign.image
+
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import com.timothyolt.evolutionarydesign.R
+import kotlinx.coroutines.*
+import java.io.BufferedInputStream
+import java.net.HttpURLConnection
+import java.net.URL
+
+class ImageViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+    constructor(parent: ViewGroup) : this(
+        LayoutInflater.from(parent.context).inflate(R.layout.item_image, parent, false)
+    )
+
+    private val coroutineScope = CoroutineScope(Dispatchers.Main + SupervisorJob())
+
+    fun bind(image: Image) {
+        itemView.findViewById<TextView>(R.id.imageDescription).text = image.title
+        val imageView = itemView.findViewById<ImageView>(R.id.image)
+        imageView.contentDescription = image.description
+        coroutineScope.launch {
+            val bitmap = getBitmap(image.link)
+            imageView.setImageBitmap(bitmap)
+        }
+    }
+
+    private suspend fun getBitmap(imageUrl: String): Bitmap = withContext(Dispatchers.IO) {
+        val connection = URL(imageUrl)
+            .let { it.openConnection() as HttpURLConnection }
+            .apply { requestMethod = "GET" }
+
+        val inputStream = BufferedInputStream(connection.inputStream)
+
+        val data = inputStream.use { inputStream.readBytes() }
+
+        connection.disconnect()
```

Comment:
I feel like this is becoming a pattern. Do you think it hurts our example to extract an extension function like:

```kotlin
fun HttpURLConnection.readBytes(): ByteArray {
    try {
    return BufferedInputStream(inputStream)
        .use { readBytes() }
    } catch(e: Throwable) {
        throw NetworkException(e)
    } finally {
        disconnect()
    }
}
```

Reply:
Good call, I would want to include the openConnection in this function though for symmetry and better cohesion with the disconnect.

Reply:
So I learned a new thing just now. It's a good idea to _not_ disconnect an HttpURLConnection if you already closed the input stream.
https://docs.oracle.com/javase/6/docs/technotes/guides/net/http-keepalive.html